sql/analyzer/resolve_columns.go

```diff
@@ -183,34 +236,524 @@ func dedupStrings(in []string) []string {
 
 // qualifyColumns assigns a table to any column expressions that don't have one already
 func qualifyColumns(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
+	var nestingLevel int
+	symbols := make(availableNames)
 	return transform.Node(n, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
 		if _, ok := n.(sql.Expressioner); !ok || n.Resolved() {
 			return n, transform.SameTree, nil
 		}
 
-		symbols := getNodeAvailableNames(n, scope)
+		symbols = getNodeAvailableNames(n, scope, symbols, nestingLevel)
+		nestingLevel++
 
-		return transform.OneNodeExprsWithNode(n, func(_ sql.Node, e sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
+		return transform.OneNodeExprsWithNode(n, func(n sql.Node, e sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
 			return qualifyExpression(e, symbols)
 		})
 	})
 }
 
+type scope struct {
```

Comment:
Unfortunate overloading, maybe call this namespace

Reply:
Seems like this and other new analysis types, related methods belong in a new file, symbol_resolution.go or similar